File: 2nd_quarter/week06/Whaeun/당구연습.cpp
#include <string>
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

vector<int> solution(int m, int n, int startX, int startY, vector<vector<int>> balls) {
    vector<int> answer;
    
    for(int i = 0; i < balls.size(); i++){
        int k = 100000000;
        
        if (startX != balls[i][0] || balls[i][1] > startY){
            if( k > pow((startX - balls[i][0]), 2) + pow((startY + balls[i][1]), 2)){
                k = pow((startX - balls[i][0]), 2) + pow((startY + balls[i][1]), 2);
            }
        }
        if(startY != balls[i][1] || balls[i][0] > startX){
            if( k > pow((startX + balls[i][0]), 2) + pow((startY - balls[i][1]), 2)){
                k = pow((startX + balls[i][0]), 2) + pow((startY - balls[i][1]), 2);
            }
        }
        if(startX != balls[i][0] || balls[i][1] < startY){
            if( k > pow((startX - balls[i][0]), 2) + pow((2 * n - balls[i][1] - startY), 2)){
                k = pow((startX - balls[i][0]), 2) + pow((2 * n - balls[i][1] - startY), 2);
            }
        }
        if(startY != balls[i][1] || balls[i][0] < startX){
            if( k > pow((2 * m - balls[i][0] - startX), 2) + pow((startY-balls[i][1]), 2)){
                k = pow((2 * m - balls[i][0] - startX), 2) + pow((startY-balls[i][1]), 2);
            }
        }
            
        answer.push_back(k);
            
    }
                    
    
    return answer;
}

File: 2nd_quarter/week06/Rash/당구연습.py
# 풀이 게획
# 1. 같은 x축에 있다면 무조건 y축을 이용해야한다.  y축 반전 두개만 구하고 그중에서 최솟값을 구하면 된다.
# 2. 같은 y축에 있다면 무조건 x축을 이용해야한다.  x축 반전 두개만 구하고 그중에서 최솟값을 구한다.
# 3. 만약 둘다 동일선상에 없다면 1,2,3,4 전부 구하고 최솟값을 구하면 되지 않을까?


import math


def solution(m, n, startX, startY, balls):
    answer = []
    for i in range(len(balls)):
        X = startX
        Y = startY

        Top = (balls[i][0] - X) ** 2 + (balls[i][1] - (Y + 2 * (n - Y))) ** 2
        Bottom = (balls[i][0] - X) ** 2 + (balls[i][1] + Y) ** 2
        Left = (balls[i][0] + X) ** 2 + (balls[i][1] - Y) ** 2
        Right = (balls[i][0] - (X + 2 * (m - X))) ** 2 + (balls[i][1] - Y) ** 2

        if startX == balls[i][0] and startY > balls[i][1]:
            distance = min(Left, Right, Top)

        elif startX == balls[i][0] and startY < balls[i][1]:
            distance = min(Left, Right, Bottom)

        elif startY == balls[i][1] and startX > balls[i][0]:
            distance = min(Top, Bottom, Right)

        elif startY == balls[i][1] and startX < balls[i][0]:
            distance = min(Top, Bottom, Left)

        else:
            distance = min(Top, Bottom, Left, Right)

        answer.append(distance)

    return answer
